panacea/query_client.go

```diff
@@ -0,0 +1,105 @@
+package panacea
+
+import (
+	"context"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	"github.com/medibloc/panacea-core/v2/types/compkey"
+	aoltypes "github.com/medibloc/panacea-core/v2/x/aol/types"
+)
+
+type QueryClient struct {
+	rpcClient         *RpcClient
+	interfaceRegistry codectypes.InterfaceRegistry
+}
+
+// NewQueryClient set QueryClient with rpcClient & and returns, if successful,
+// a QueryClient that can be used to add query function.
+func NewQueryClient(ctx context.Context, chainID, rpcAddr string, trustedBlockHeight int, trustedBlockHash []byte) (*QueryClient, error) {
+	rpcClient, err := NewRpcClient(ctx, chainID, rpcAddr, trustedBlockHeight, trustedBlockHash)
+	if err != nil {
+		return nil, err
+	}
+
+	return &QueryClient{
+		rpcClient:         rpcClient,
+		interfaceRegistry: makeInterfaceRegistry(),
+	}, nil
+}
+
+// Below are examples of query function that use GetStoreData function to verify queried result.
+// Need to set storeKey and key inside the query function, and change type to expected type.
+
+// GetAccount returns account from address.
+func (c QueryClient) GetAccount(address string) (authtypes.AccountI, error) {
+	acc, err := sdk.GetFromBech32(address, "panacea")
+	if err != nil {
+		return nil, err
+	}
+
+	key := authtypes.AddressStoreKey(acc)
+	bz, err := c.rpcClient.GetStoreData(context.Background(), authtypes.StoreKey, key)
+	if err != nil {
+		return nil, err
+	}
+
+	var accountAny codectypes.Any
+	err = accountAny.Unmarshal(bz)
+	if err != nil {
+		return nil, err
+	}
+
+	var account authtypes.AccountI
+	err = c.interfaceRegistry.UnpackAny(&accountAny, &account)
+	if err != nil {
+		return nil, err
+	}
+
+	return account, nil
+}
+
+// GetBalance returns balance from address.
+func (c QueryClient) GetBalance(address string) (sdk.Coin, error) {
+	acc, err := sdk.GetFromBech32(address, "panacea")
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	var denom = "umed"
```

Comment:
It would be better to subtract this value as const or receive it as a parameter.

Reply:
Thanks, I reflect in a new commit.